python-qcgrids/qcgrids/ext.pyx

```diff
@@ -157,3 +167,386 @@ cdef class Cellgrid(object):
+        while np.PyArray_MultiIter_NOTDONE(it):
+            (<double*>np.PyArray_MultiIter_DATA(it, 1))[0] = \
+                deref(self._this).deriv_inv(deref(<double*>np.PyArray_MultiIter_DATA(it, 0)))
+            np.PyArray_MultiIter_NEXT(it)
+        return result if isinstance(x, np.ndarray) else result[()]
+
+    def deriv2(self, x):
+        """Compute the second derivative. Works as u-func."""
+        result = np.zeros_like(x, dtype=float)
+        cdef np.broadcast it = np.broadcast(np.asarray(x, dtype=float), result)
+        while np.PyArray_MultiIter_NOTDONE(it):
+            (<double*>np.PyArray_MultiIter_DATA(it, 1))[0] = \
+                deref(self._this).deriv2(deref(<double*>np.PyArray_MultiIter_DATA(it, 0)))
+            np.PyArray_MultiIter_NEXT(it)
+        return result if isinstance(x, np.ndarray) else result[()]
+
+    def deriv2_inv(self, x):
+        """Compute the second derivative of the inverse. Works as u-func."""
+        result = np.zeros_like(x, dtype=float)
+        cdef np.broadcast it = np.broadcast(np.asarray(x, dtype=float), result)
+        while np.PyArray_MultiIter_NOTDONE(it):
+            (<double*>np.PyArray_MultiIter_DATA(it, 1))[0] = \
+                deref(self._this).deriv2_inv(deref(<double*>np.PyArray_MultiIter_DATA(it, 0)))
+            np.PyArray_MultiIter_NEXT(it)
+        return result if isinstance(x, np.ndarray) else result[()]
+
+
+cdef class Exp(ScalarFunction):
+    """Exponential function: y = exp(slope*x + offset)."""
+
+    def __cinit__(self, double slope, double offset):
+        self._this.reset(new scalarfns.Exp(slope, offset))
+
+    def __init__(self, double slope, double offset):
+        pass
+
+    property slope:
+        def __get__(self):
+            return deref(<scalarfns.Exp*?>self._this.get()).slope()
+
+    property offset:
+        def __get__(self):
+            return deref(<scalarfns.Exp*?>self._this.get()).offset()
+
+
+cdef class Ln(ScalarFunction):
+    """Logarithmic function: y = prefac*ln(alpha*x)."""
+
+    def __cinit__(self, double prefac, double alpha):
+        self._this.reset(new scalarfns.Ln(prefac, alpha))
+
+    def __init__(self, double prefac, double alpha):
+        pass
+
+    property prefac:
+        def __get__(self):
+            return deref(<scalarfns.Ln*?>self._this.get()).prefac()
+
+    property alpha:
+        def __get__(self):
+            return deref(<scalarfns.Ln*?>self._this.get()).alpha()
+
+
+cdef class Linear(ScalarFunction):
+    """Linear function: y = slone*x + offset."""
+
+    def __cinit__(self, double slope, double offset):
+        self._this.reset(new scalarfns.Linear(slope, offset))
+
+    def __init__(self, double slope, double offset):
+        pass
+
+    property slope:
+        def __get__(self):
+            return deref(<scalarfns.Linear*?>self._this.get()).slope()
+
+    property offset:
+        def __get__(self):
+            return deref(<scalarfns.Linear*?>self._this.get()).offset()
```

Comment:
Also stuff like this

Reply:
The question mark is for type-checked casting. See http://cython.readthedocs.io/en/latest/src/reference/language_basics.html?highlight=casting#type-checking

I know it looks ugly but this seems to be the normal way to wrap access to a subclass member. Any suggestions for improvements are obviously welcome.